denokv/main.rs

```diff
@@ -97,8 +248,89 @@ async fn main() -> Result<(), anyhow::Error> {
   Ok(())
 }
 
-fn open_sqlite(path: &Path) -> Result<(Sqlite, String), anyhow::Error> {
-  let conn = Connection::open(path)?;
+async fn run_sync(
+  config: &Config,
+  options: &ReplicaOptions,
+  continuous: bool,
+  initial_sync_ok_tx: Option<oneshot::Sender<()>>,
+) -> anyhow::Result<()> {
+  let mut s3_config = aws_config::from_env()
+    .sleep_impl(Arc::new(TokioSleep::new()))
+    .retry_config(RetryConfig::standard().with_max_attempts(std::u32::MAX));
+
+  if let Some(endpoint) = &options.s3_endpoint {
+    s3_config = s3_config.endpoint_url(endpoint);
+  }
+
+  let https_proxy = env::var("https_proxy")
+    .or_else(|_| env::var("HTTPS_PROXY"))
+    .ok();
+  if let Some(https_proxy) = https_proxy {
+    let proxy = {
+      let proxy_uri = https_proxy.parse().unwrap();
+      let proxy = Proxy::new(Intercept::All, proxy_uri);
+      let connector = HttpConnector::new();
+      ProxyConnector::from_proxy(connector, proxy).unwrap()
+    };
+    let hyper_client =
+      aws_smithy_client::hyper_ext::Adapter::builder().build(proxy);
+
+    s3_config = s3_config.http_connector(hyper_client);
+  }
+
+  let s3_config = s3_config.load().await;
+  let s3_client = aws_sdk_s3::Client::new(&s3_config);
+
+  let db = rusqlite::Connection::open(&config.sqlite_path)?;
+  let mut ttc = TimeTravelControl::open(db)?;
+  let s3_config = DatabaseBackupSourceS3Config {
+    bucket: options
+      .s3_bucket
+      .clone()
+      .ok_or_else(|| anyhow::anyhow!("--s3-bucket not set"))?,
+    prefix: options.s3_prefix.clone().unwrap_or_default(),
+  };
+  if !s3_config.prefix.ends_with('/') {
+    anyhow::bail!("--s3-prefix must end with a slash")
+  }
+
+  ttc.init_s3(&s3_config)?;
+  let source = DatabaseBackupSourceS3::new(s3_client, s3_config);
+  ttc.ensure_initial_snapshot_completed(&source).await?;
+
+  log::info!("Initial snapshot is complete, starting sync.");
+  ttc.checkout([0xffu8; 10])?;
+  drop(initial_sync_ok_tx);
+
+  loop {
+    ttc.sync(&source).await?;
+
+    if !continuous {
+      return Ok(());
+    }
+
+    // In continuous mode, always advance snapshot to latest version
+    ttc.checkout([0xffu8; 10])?;
+
+    let sleep_duration = std::time::Duration::from_millis(
+      10000 + rand::thread_rng().gen_range(0..5000),
+    );
+    tokio::time::sleep(sleep_duration).await;
+  }
+}
+
+fn open_sqlite(
+  path: &Path,
+  read_only: bool,
+) -> Result<(Sqlite, String), anyhow::Error> {
+  let flags = if read_only {
+    OpenFlags::SQLITE_OPEN_READ_ONLY
+      | OpenFlags::SQLITE_OPEN_NO_MUTEX
+      | OpenFlags::SQLITE_OPEN_URI
```

Comment:
Why `SQLITE_OPEN_URI`? I really don't like this flag.

Reply:
Because that's in `OpenFlags::default()`

Reply:
Can you remove it? We should do all URI parsing upfront.